internal/data.go

```diff
@@ -0,0 +1,136 @@
+package internal
+
+import (
+	"time"
+
+	api "github.com/tigrisdata/tigrisdb/api/server/v1"
+	"github.com/ugorji/go/codec"
+	"google.golang.org/grpc/codes"
+)
+
+var (
+	bh codec.BincHandle
+)
+
+type DataType byte
+
+// Note: Do not change the order. Order is important because encoder is adding the type as the first byte.
+const (
+	TableDataType DataType = iota + 1
+	CDCDataType
+	DictEncDataType
+	RawDataType
+)
+
+type Data interface {
+	GetDataType() DataType
+}
+
+func NewTableData(data []byte) Data {
+	return &TableData{
+		Ts: &Timestamp{
+			Nano: time.Now().UnixNano(),
+		},
+		UserData: data,
+	}
+}
+
+func (x *TableData) GetDataType() DataType {
+	return TableDataType
+}
+
+func (x *CDCData) GetDataType() DataType {
+	return CDCDataType
+}
+
+func NewDictEncData(encValue uint32) Data {
+	return &DictEncData{
+		Ts: &Timestamp{
+			Nano: time.Now().UnixNano(),
+		},
+		EncValue: encValue,
+	}
+}
+
+func (x *DictEncData) GetDataType() DataType {
+	return DictEncDataType
+}
+
+func NewRawData(data []byte) Data {
+	return &RawData{
+		Ts: &Timestamp{
+			Nano: time.Now().UnixNano(),
+		},
+		Data: data,
+	}
+}
+
+func (x *RawData) GetDataType() DataType {
+	return RawDataType
+}
+
+func Encode(data Data) ([]byte, error) {
+	var b []byte
+	enc := codec.NewEncoderBytes(&b, &bh)
+
+	switch data.GetDataType() {
```

Comment:
What if we just pass interface to this function and:
```
switch t := data.(type) {
case *TableData:
```
So we don't need to implement any interfaces.

Reply:
if we don't implement the interface or enforce it then the problem would be that any object will fulfill the data interface and can be passed around.

Reply:
But then encoding will fail.

Reply:
but that will be runtime, if we enforce then we can detect at compile time.